modules/caddyhttp/reverseproxy/selectionpolicies.go

```diff
@@ -78,6 +79,108 @@ func (r *RandomSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
 	return nil
 }
 
+// WeightedRoundRobinSelection is a policy that selects
+// a host based on weighted round-robin ordering.
+type WeightedRoundRobinSelection struct {
+	// The weight of each upstream in order.
+	Weights     []int `json:"weights,omitempty"`
+	index       uint32
+	totalWeight int
+}
+
+// CaddyModule returns the Caddy module information.
+func (WeightedRoundRobinSelection) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID: "http.reverse_proxy.selection_policies.weighted_round_robin",
+		New: func() caddy.Module {
+			return new(WeightedRoundRobinSelection)
+		},
+	}
+}
+
+// UnmarshalCaddyfile sets up the module from Caddyfile tokens.
+func (r *WeightedRoundRobinSelection) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+	for d.Next() {
+		args := d.RemainingArgs()
+		if len(args) == 0 {
+			return d.ArgErr()
+		}
+
+		for _, weight := range args {
+			weightInt, err := strconv.Atoi(weight)
+			if err != nil {
+				return d.Errf("invalid weight value '%s': %v", weight, err)
+			}
+			if weightInt < 1 {
+				return d.Errf("invalid weight value '%s': weight should be non-zero and positive", weight)
+			}
+			r.totalWeight += weightInt
+			r.Weights = append(r.Weights, weightInt)
+		}
+	}
+	return nil
+}
+
+// Provision sets up r.
+func (r *WeightedRoundRobinSelection) Provision(ctx caddy.Context) error {
+	if len(r.Weights) != 0 {
+		r.totalWeight = 0
+		for _, weight := range r.Weights {
+			r.totalWeight += weight
+		}
+	}
+	return nil
+}
+
+// Validate ensures that r's configuration is valid.
+func (r WeightedRoundRobinSelection) Validate() error {
+	if len(r.Weights) != 0 {
+		totalWeight := 0
+		for _, weight := range r.Weights {
+			totalWeight += weight
+		}
+		if r.totalWeight != totalWeight {
+			return fmt.Errorf("total weight miscalculated")
+		}
+	}
+	return nil
+}
+
+// Select returns an available host, if any.
+func (r *WeightedRoundRobinSelection) Select(pool UpstreamPool, _ *http.Request, _ http.ResponseWriter) *Upstream {
+	if len(pool) == 0 {
+		return nil
+	}
+	if len(r.Weights) < 2 {
+		return pool[0]
+	}
+	index := 0
+	totalWeight := 0
```

Comment:
```suggestion
	var index, totalWeight int
```

Totally just a style nit, but hear me out :)